Simple Service to make a GraphViz graph of the modules runtime dependencies

fwyzard opened the pull request:
Simple Service to make a GraphViz graph of the modules runtime dependencies:
  - draw hard dependencies according to the "consumes" dependencies;
  - draw soft dependencies to reflect the order of scheduled modue in each path;
  - draw SubProcesses in subgraphs.

Use GraphViz dot to generate an SVG representation of the dependencies:
```
dot -v -Tsvg dependency.gv -o dependency.svg
```

fwyzard committed:
DependencyGraph: simple Service to make a GraphViz graph of the modules runtime dependencies

fwyzard committed:
DependencyGraph: directly use the module ids

fwyzard committed:
DependencyGraph: improve handling of the Source module

fwyzard committed:
DependencyGraph: implement support for SubProcesses

fwyzard committed:
DependencyGraph: configure output file name

fwyzard committed:
DependencyGraph: draw SubProcesses in subgraphs

fwyzard committed:
DependencyGraph: draw soft dependencies to reflect the order of scheduled modue in each path

fwyzard committed:
DependencyGraph: add comment

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for CMSSW_9_0_X.

It involves the following packages:

FWCore/Services


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

cmsbuild commented:
+1
Tested at: 079c0892503682992b697f83f5a61c118963125b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17417/17622/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17417/17622/summary.html

Dr15Jones commented on FWCore/Services/plugins/BuildFile.xml:
```diff
@@ -1,6 +1,8 @@
 <use   name="gcc-atomic"/>
 <library   file="*.cc" name="FWCoreServicesPlugins">
   <flags   EDM_PLUGIN="1"/>
+  <flags   CXXFLAGS="-Wno-maybe-uninitialized"/>
```
@smuzaffar are you OK with this flag?

fwyzard replied:
if there is a way f disabling the check only in the boost headers, that works for me as well...

smuzaffar replied:
@fwyzard , no there is no way to disable it only for boost. If you are worrying about getting these warnings in IB logs then we do ignore warnings coming from externals.

fwyzard replied:
The problem is that we have `-Werror=maybe-uninitialized` so they become actual build errors.
 If you prefer I can use `-Wno-error=maybe-uninitialized` to turn them back into warnings.

fwyzard committed:
DependencyGraph: move to a separate plugin

cmsbuild commented:
Pull request #17417 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

fwyzard committed:
DependencyGraph: use explicit information about the Source module

cmsbuild commented:
Pull request #17417 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

fwyzard commented:
OK, I have updated the service based on the feedback at the core software meeting:
  - moved it to a different plugin, so the warnings can be disabled only for this service
  - gather information about the Source module explicitly

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented on FWCore/Services/plugins/BuildFile.xml:
```diff
@@ -1,6 +1,12 @@
+<library   file="DependencyGraph.cc" name="FWCoreServicesDependencyGraphPlugin">
+  <flags   EDM_PLUGIN="1"/>
+  <flags   CXXFLAGS="-Wno-maybe-uninitialized"/>
+</library>
```
@Dr15Jones , this looks good to me.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

davidlange6 commented:
apologies that I missed the discussion - what warnings are we trying to work around in this PR? 
(ok, I see I have some invalid tool warnings to clean up...but thats not the issue)

fwyzard commented:
Boost has some "maybe-uninitialized" variables, and since we normally compile with `-Werror=maybe-uninitialized` those make the compilation fail.
My workaround was to add `-Wno-maybe-uninitialized` to the BuildFile, and Shahzad asked to split the plugin so that the flag could be applied only to the new service.

davidlange6 commented:
ok , thanks - hopefully a new version of boost fixes such issues:)

davidlange6 commented:
+1

cmsbuild commented:
+1
Tested at: 1c6d4842b71e0370c56d49ce82a28c3d479d26b9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17417/17696/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17417/17696/summary.html
